### DISCUSSION TITLE
Support for uuid primary keys

### DISCUSSION THREAD

**Discussion by SrNetoChan** [koordinates/kart#874] [Ideas]
I have a use case, where the data model build on PostgreSQL/PostGIS uses the uuid-ossp extension to populate primary keys on all tables. When Import this tables to kart, all this fields are converted into text.

I have tried to convert them back to the uuid datatype, which works fine, except that (related or not) from this point on wards I am no longer able to view the geometric changes on the qgis plugin.

I wonder if it would be possible to add support to uuid datatypes without casting it to text.

**Comment by SrNetoChan** [comment]
Actually, even before I convert the column to uui datatype I get the following error when trying to see the changes:

{
  "ExceptionDetails": {
    "Type": "KeyError",
    "Message": "'identificador'"
  },
  "Environment": {
    "Qgis Version": "3.22.16-Bia\u0142owie\u017ca",
    "Operating System": "linux",
    "Locale": "en"
  },
  "Trace": [
    {
      "Name": "treeItemChanged",
      "Filename": "diffviewer.py",
      "LineNo": 186,
      "Variables": {
        "self": "<kart.gui.diffviewer.DiffViewerWidget object at 0x7f0dca0fde10>",
        "current": "<kart.gui.diffviewer.FeatureItem object at 0x7f0dca1255a0>",
        "previous": "None"
      }
    },
    {
      "Name": "_createLayers",
      "Filename": "diffviewer.py",
      "LineNo": 481,
      "Variables": {
        "self": "<kart.gui.diffviewer.DiffViewerWidget object at 0x7f0dca0fde10>"
      }
    },
    {
      "Name": "_createFeatureDiffLayers",
      "Filename": "diffviewer.py",
      "LineNo": 528,
      "Variables": {
        "self": "<kart.gui.diffviewer.DiffViewerWidget object at 0x7f0dca0fde10>",
        "old": "{}",
        "new": "{'type': 'Feature', 'geometry': {'type': 'LineString', 'coordinates': [[-101793.21560871392, -98383.80476362782, 0.0], [-102241.77129073141, -88366.06119857082, 0.0], [-93569.69477172685, -82086.28165032616, 0.0], [-84748.09969204979, -81189.17028629119, 0.0], [-83701.46976734234, -86870.8755918459, 0.0], [-80113.02431120252, -78946.39187620381, 0.0], [-76524.5788550627, -72816.13088863163, 0.0]]}, 'properties': {'identificador': '234234234234', 'inicio_objeto': None, 'fim_objeto': None, 'valor_tipo_curva': None}, 'id': 'curva_de_nivel:feature:234234234234:I'}",
        "dataset": "curva_de_nivel",
        "layer": "<QgsVectorLayer: 'new' (memory)>",
        "crs": "EPSG:3763",
        "idField": "identificador",
        "ref": "{'type': 'Feature', 'geometry': {'type': 'LineString', 'coordinates': [[-101793.21560871392, -98383.80476362782, 0.0], [-102241.77129073141, -88366.06119857082, 0.0], [-93569.69477172685, -82086.28165032616, 0.0], [-84748.09969204979, -81189.17028629119, 0.0], [-83701.46976734234, -86870.8755918459, 0.0], [-80113.02431120252, -78946.39187620381, 0.0], [-76524.5788550627, -72816.13088863163, 0.0]]}, 'properties': {'identificador': '234234234234', 'inicio_objeto': None, 'fim_objeto': None, 'valor_tipo_curva': None}, 'id': 'curva_de_nivel:feature:234234234234:I'}",
        "refGeom": "{'type': 'LineString', 'coordinates': [[-101793.21560871392, -98383.80476362782, 0.0], [-102241.77129073141, -88366.06119857082, 0.0], [-93569.69477172685, -82086.28165032616, 0.0], [-84748.09969204979, -81189.17028629119, 0.0], [-83701.46976734234, -86870.8755918459, 0.0], [-80113.02431120252, -78946.39187620381, 0.0], [-76524.5788550627, -72816.13088863163, 0.0]]}",
        "options": "<qgis._core.QgsVectorLayer.LayerOptions object at 0x7f0dca125d80>",
        "geomtype": "LineString",
        "geoms": "[]",
        "feat": "{'type': 'Feature', 'geometry': {'type': 'LineString', 'coordinates': [[-101793.21560871392, -98383.80476362782, 0.0], [-102241.77129073141, -88366.06119857082, 0.0], [-93569.69477172685, -82086.28165032616, 0.0], [-84748.09969204979, -81189.17028629119, 0.0], [-83701.46976734234, -86870.8755918459, 0.0], [-80113.02431120252, -78946.39187620381, 0.0], [-76524.5788550627, -72816.13088863163, 0.0]]}, 'properties': {'identificador': '234234234234', 'inicio_objeto': None, 'fim_objeto': None, 'valor_tipo_curva': None}, 'id': 'curva_de_nivel:feature:234234234234:I'}",
        "geom": "<QgsGeometry: LineStringZ (-101793.21560871391557157 -98383.80476362782064825 0, -102241.77129073141259141 -88366.06119857082376257 0, -93569.69477172684855759 -82086.28165032615652308 0, -84748.09969204978551716 -81189.17028629119158722 0, -83701.4697673423361266 -86870.875591845906456 0, -80113.02431120252003893 -78946.39187620380835142 0, -76524.57885506270395126 -72816.13088863162556663 0)>",
        "props": "{'identificador': '234234234234', 'inicio_objeto': None, 'fim_objeto': None, 'valor_tipo_curva': None}",
        "feature": "<qgis._core.QgsFeature object at 0x7f0dca126050>"
      }
    }
  ]
}

Maybe I should open a ticket on the plugin?

**Comment by rcoup** [comment]
@SrNetoChan using a text field as a dataset primary key should work, anywhere it doesn't is a bug. Most datasets have integer PKs so it's possible a regression has happened.

> I have tried to convert them back to the uuid datatype, which works fine, except that (related or not) from this point on wards I am no longer able to view the geometric changes on the qgis plugin.

So you're doing something like `ALTER TABLE kart.mydataset ALTER COLUMN id TYPE uuid USING (uuid(id));`?

**Reply by rcoup** [reply]
@olsen232 natively supporting UUID PKs could be a fairly elegant way to sidestep the complexity of merging integer PKs?

**Reply by SrNetoChan** [reply]
In that case it seems it's a bug on the plugin. The error I mention happens when I hope the difference viewer even before I convert the primary key field into uuid.

BTW I am using:

ALTER TABLE my_table
ALTER COLUMN identificador type uuid using identificador::uuid;
